pints/_transformation.py

```diff
@@ -0,0 +1,1120 @@
+        from the inverse function theorem, i.e. the matrix inverse of the
+        Jacobian matrix of an invertible function is the Jacobian matrix of the
+        inverse function.
+
+        To transform the provided standard deviation :math:`\boldsymbol{s}`, we
+        assume the covariance matrix :math:`\mathbf{C}(\boldsymbol{p})` above
+        is a diagonal matrix with :math:`\boldsymbol{s}^2` on the diagonal,
+        such that
+
+        .. math::
+            s_i(\boldsymbol{q}) =
+                \left(
+                    \mathbf{J}^{-1} (\mathbf{J}^{-1})^T
+                \right)^{1/2}_{i, i}
+                s_i(\boldsymbol{p}).
+
+        References
+        ----------
+        .. [1] How to Obtain Those Nasty Standard Errors From Transformed Data
+               Erik Jorgensen and Asger Roer Pedersen,
+               http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.47.9023
+        """
+        jac_inv = np.linalg.pinv(self.jacobian(q))
+        return s * np.sqrt(np.diagonal(np.matmul(jac_inv, jac_inv.T)))
+
+    def jacobian(self, q):
+        r"""
+        Returns the Jacobian matrix of the transformation calculated at the
+        parameter vector ``q`` in the search space. For a transformation
+        :math:`\boldsymbol{q} = \boldsymbol{f}(\boldsymbol{p})`, the Jacobian
+        matrix is defined as
+
+        .. math::
+            \mathbf{J} =
+                \left[\frac{\partial \boldsymbol{f}^{-1}}{\partial q_1} \quad
+                 \frac{\partial \boldsymbol{f}^{-1}}{\partial q_2} \quad
+                 \cdots \right].
+
+        *This is an optional method.* It is needed when transformation of
+        standard deviation :meth:`Transformation.convert_standard_deviation` or
+        covariance matrix :meth:`Transformation.convert_covariance_matrix` is
+        needed, or when ``evaluateS1()`` is needed.
+        """
+        raise NotImplementedError
+
+    def jacobian_S1(self, q):
+        r"""
+        Computes the Jacobian matrix of the transformation calculated at the
+        parameter vector ``q`` in the search space, and returns the result
+        along with the partial derivatives of the result with respect to the
+        parameters.
+
+        The returned data is a tuple ``(S, S')`` where ``S`` is a
+        ``n_parameters`` by ``n_parameters`` matrix and ``S'`` is a sequence of
+        ``n_parameters`` matrices.
+
+        *This is an optional method.* It is needed when the transformation is
+        used along with a non-element-wise transformation in
+        :class:`ComposedTransformation`.
+        """
+        raise NotImplementedError
+
+    def log_jacobian_det(self, q):
+        """
+        Returns the logarithm of the absolute value of the determinant of the
+        Jacobian matrix of the transformation :meth:`Transformation.jacobian`
+        calculated at the parameter vector ``q`` in the search space.
+
+        *This is an optional method.* It is needed when transformation is
```

Comment:
Is the implementation below always suitable? Maybe the "optional" text should mention its implemented?
Minor detail :D

Reply:
Something like *This method only works if the optional method X() is implemented"

Reply:
And the default implementation could be explained in the docstring?

Reply:
Changed it to
```
        The default implementation numerically calculates the determinant of    
        the full matrix. If there is an analytic expression for the specific    
        transformation, a reimplementation of this method may be preferred.     
                                                                                
        *This is an optional method that only works if the optional method      
        :meth:`Transformation.jacobian` is implemented.* It is needed when      
        transformation is performed on :class:`LogPDF` and/or that requires     
        ``evaluateS1()``; e.g. not necessary if it's used for                   
        :class:`ErrorMeasure` without :meth:`ErrorMeasure.evaluateS1()`.
```
Does it explain better? Any suggestion?